### PR TITLE
Reduce requests

### PR DESCRIPTION
**Pull request by gshokanov**
This pull request aims to reduce client calls to server. I had to rewrite server middleware to fully control what the server sends.

### CODE REVIEW

File: client/src/shared/services/documentService.js [hunk 1]
```diff
@@ -6,14 +6,18 @@ export default angular.module('shared')
         function getPopularDocuments() {
             return $http.get(`${SERVER_URL}/documents/popular`);
         }
-        function searchDocuments(query, filtersArray) {
+        function searchDocuments(query, filtersArray, countFilters) {
```
Comment (Olga1984) [inline client/src/shared/services/documentService.js]: Please, rename variable to isCountFilters to underlain boolean or something like this.

File: client/src/shared/services/documentService.js [hunk 2]
```diff
@@ -6,14 +6,18 @@ export default angular.module('shared')
...
             let filterStr = '';
+            let filtersCount = '';
             if (filtersArray) {
                 filterStr = filtersArray.reduce((acc, filterId) => (
                     `${acc}&filterId=${filterId}`
                 ), '');
             }
-            return $http.get(`${SERVER_URL}/documents?search=${query}${filterStr}`);
+            if (countFilters === true) {
```
Comment (Olga1984) [inline client/src/shared/services/documentService.js]: the same

File: client/src/searchPage/searchPageController.js [hunk 3]
```diff
@@ -28,22 +30,50 @@ export default angular.module('searchPage')
...
-                        this.filters = filters;
-                        const filtersArr = (filters && filters.split(',').map(Number)) || [];
-                        searchDocuments(query, filtersArr);
+                    const { query: newQuery, filters: newFilters } = params;
+                    if (this.query !== newQuery || this.filters !== newFilters) {
+                        this.filters = newFilters;
+                        this.filtersArr = parseFilters(this.filters);
+                        const requestFiltersCount = this.query !== newQuery;
+                        this.query = newQuery;
+                        searchDocuments(newQuery, this.filtersArr, requestFiltersCount);
```
Comment (Olga1984) [inline client/src/searchPage/searchPageController.js]: the same for requestFiltersCount

File: client/src/searchPage/searchPageController.js [hunk 4]
```diff
@@ -28,22 +30,50 @@ export default angular.module('searchPage')
...
-                    if (this.query !== query || this.filters !== filters) {
-                        this.query = query;
-                        this.filters = filters;
-                        const filtersArr = (filters && filters.split(',').map(Number)) || [];
-                        searchDocuments(query, filtersArr);
+                    const { query: newQuery, filters: newFilters } = params;
+                    if (this.query !== newQuery || this.filters !== newFilters) {
+                        this.filters = newFilters;
+                        this.filtersArr = parseFilters(this.filters);
+                        const requestFiltersCount = this.query !== newQuery;
```
Comment (Olga1984) [inline client/src/searchPage/searchPageController.js]: the same for requestFiltersCount

Comment (Olga1984) [reply client/src/searchPage/searchPageController.js]: something like is or shouldRequestFiltersCount or shouldGetFiltersCount, something more talking about the goal and showing that it is a boolean

File: server/server.js [hunk 5]
```diff
@@ -18,36 +18,47 @@ server.use(middlewares);
...
-    if (filtersCount !== 'true') {
-        return res.sendStatus(400);
+    const result = {
+        documents: [],
+    };
+    const regex = new RegExp(search, 'i');
+    const filtersArr = Array.isArray(filterId)
+    ? filterId.map(Number)
+    : [Number(filterId)];
+    if (filtersCount === 'true') {
```
Comment (Olga1984) [inline server/server.js]: the same

File: server/server.js [hunk 6]
```diff
@@ -18,36 +18,47 @@ server.use(middlewares);
...
+    documents.forEach((document) => {
+        const { filterId: documentFilter, headline, text } = document;
         if (!regex.test(headline) && !regex.test(text)) {
             return;
         }
-        if (cache[filterId]) {
-            cache[filterId] += 1;
-        } else {
-            cache[filterId] = 1;
+        if (filtersCount === 'true') {
```
Comment (Olga1984) [inline server/server.js]: the same

File: server/server.js [hunk 7]
```diff
@@ -18,36 +18,47 @@ server.use(middlewares);
...
         }
+        if (filterId && !filtersArr.includes(documentFilter)) {
+            return;
+        }
+        result.documents.push(document);
     });
-    res.jsonp(cache);
-})
+    if (result.documents.length === 0) {
+        if (filtersCount === 'true') {
```
Comment (Olga1984) [inline server/server.js]: the same

File: server/server.js [hunk 8]
```diff
@@ -18,36 +18,47 @@ server.use(middlewares);
...
     "/api/*": "/$1"
 }));
 
 
-server.use((req, res, next) => {
+server.use('/documents', (req, res, next) => {
     const { documents } = data;
-    const { q: searchQuery, filtersCount } = req.query;
-    if (typeof filtersCount !== 'string') {
+    const { search, filterId, filtersCount } = req.query;
```
Comment (Olga1984) [inline server/server.js]: the same with boolean and searchStr or searchString for search, I think, would be better.

File: client/src/shared/services/documentService.js (hunk 1 above)
Comment (gshokanov) [reply client/src/shared/services/documentService.js]: fixed

File: client/src/shared/services/documentService.js (hunk 2 above)
Comment (gshokanov) [reply client/src/shared/services/documentService.js]: fixed

File: client/src/searchPage/searchPageController.js (hunk 3 above)
Comment (gshokanov) [reply client/src/searchPage/searchPageController.js]: fixed

File: client/src/searchPage/searchPageController.js (hunk 4 above)
Comment (gshokanov) [reply client/src/searchPage/searchPageController.js]: fixed

File: server/server.js (hunk 5 above)
Comment (gshokanov) [reply server/server.js]: fixed

File: server/server.js (hunk 6 above)
Comment (gshokanov) [reply server/server.js]: fixed

File: server/server.js (hunk 7 above)
Comment (gshokanov) [reply server/server.js]: fixed

File: server/server.js (hunk 8 above)
Comment (gshokanov) [reply server/server.js]: fixed

**Review by Olga1984 — APPROVED** [review]

**Pull request merged**